src/storage/src/hummock/cache.rs

```diff
@@ -256,47 +333,79 @@ impl<K: PartialEq + Default, T> LruCacheShard<K, T> {
+        debug_assert!(!e.is_null());
+        #[cfg(debug_assertions)]
+        {
+            assert!((*e).is_in_lru());
+            (*e).set_in_lru(false);
+        }
+
         (*(*e).next).prev = (*e).prev;
         (*(*e).prev).next = (*e).next;
         (*e).prev = null_mut();
         (*e).next = null_mut();
-        self.lru_usage -= (*e).charge;
+        self.lru_usage.fetch_sub((*e).charge, Ordering::Relaxed);
     }
 
     // insert entry in the end of the linked-list
     unsafe fn lru_insert(&mut self, e: *mut LruHandle<K, T>) {
+        debug_assert!(!e.is_null());
+        #[cfg(debug_assertions)]
+        {
+            assert!(!(*e).is_in_lru());
+            (*e).set_in_lru(true);
+        }
+
         (*e).next = self.lru.as_mut();
         (*e).prev = self.lru.prev;
         (*(*e).prev).next = e;
         (*(*e).next).prev = e;
-        self.lru_usage += (*e).charge;
+        self.lru_usage.fetch_add((*e).charge, Ordering::Relaxed);
     }
 
     unsafe fn evict_from_lru(&mut self, charge: usize, last_reference_list: &mut Vec<T>) {
-        while self.usage + charge > self.capacity && !std::ptr::eq(self.lru.next, self.lru.as_mut())
+        // TODO: may want to optimize by only loading at the beginning and storing at the end for
+        // only once.
+        while self.usage.load(Ordering::Relaxed) + charge > self.capacity
+            && !std::ptr::eq(self.lru.next, self.lru.as_mut())
         {
             let old_ptr = self.lru.next;
-            self.table.remove((*old_ptr).hash, &(*old_ptr).key);
-            assert!((*old_ptr).is_in_cache());
+            self.table.remove((*old_ptr).hash, (*old_ptr).get_key());
             self.lru_remove(old_ptr);
-            (*old_ptr).set_in_cache(false);
-            self.usage -= (*old_ptr).charge;
-            let mut node = Box::from_raw(old_ptr);
-            let data = node.value.take().unwrap();
-            last_reference_list.push(data);
-            self.recyle_handle_object(node);
+            let value = self.clear_handle(old_ptr);
+            last_reference_list.push(value);
         }
     }
 
-    fn recyle_handle_object(&mut self, mut node: Box<LruHandle<K, T>>) {
+    /// Clear a currently used handle and recycle it if possible
+    unsafe fn clear_handle(&mut self, h: *mut LruHandle<K, T>) -> T {
+        debug_assert!(!h.is_null());
+        debug_assert!((*h).kv.is_some());
+        #[cfg(debug_assertions)]
+        assert!(!(*h).is_in_lru());
+        debug_assert!(!(*h).is_in_cache());
+        debug_assert!(!(*h).has_refs());
+        self.usage.fetch_sub((*h).charge, Ordering::Relaxed);
+        let value = (*h).take_kv().1;
+        self.try_recycle_handle_object(h);
+        value
+    }
+
+    /// Try to recycle a handle object if the object pool is not full.
+    ///
+    /// The handle should already cleared its kv.
+    unsafe fn try_recycle_handle_object(&mut self, h: *mut LruHandle<K, T>) {
```

Comment:
we must run `Box::from_raw` outside `if` statement. Because if this node does not need to move to `object_pool`, it must be destruct by convert to a box pointer.

Reply:
Agreed. Thanks for pointing it out!

Reply:
Fixed